plugins/processors/k8sdecorator/stores/utils.go

```diff
@@ -51,42 +53,47 @@ func parseDeploymentFromReplicaSet(name string) string {
 		return ""
 	}
 
-	if !stringInRuneset(suffix, kubeAllowedStringAlphaNums) {
+	if containUnexpectedRuneSet(suffix, deploymentUnexpectedRegEx) {
 		// Invalid suffix
 		return ""
 	}
 
 	return name[:lastDash]
 }
 
-// get the cronJob name by stripping the last dash following some rules
-// return empty if it is not following the rule
+// Get the cronJob name by stripping the last dash following by the naming convention: JobName-UnixTime
+// based on https://github.com/kubernetes/kubernetes/blob/c4d752765b3bbac2237bf87cf0b1c2e307844666/pkg/controller/cronjob/cronjob_controllerv2.go#L594-L596.
+// Before v1.21 CronJob in Kubernetes has used Unix Time in second; after v1.21 is a Unix Time in Minutes.
+
 func parseCronJobFromJob(name string) string {
 	lastDash := strings.LastIndexAny(name, "-")
+
+	//Return empty since the naming convention is: JobName-UnixTime, if it does not have the "-", meanings the job name is empty
 	if lastDash == -1 {
-		// No dash
 		return ""
 	}
+
 	suffix := name[lastDash+1:]
-	if len(suffix) != 10 {
-		// Invalid suffix if it is not 10 rune
+	suffixInt, _ := strconv.ParseInt(suffix, 10, 64)
+	//Convert Unix Time In Minutes to Unix Time
+	suffixStringMultiply := strconv.FormatInt(suffixInt*60, 10)
+
+	//Checking if the suffix is a unix time by checking: the length and contains character
+	// - Checking for the length: CronJobControllerV2 is Unix Time in Minutes (7-9 characters) while CronJob is Unix Time (10 characters).
+	//   However, multiply by 60 to convert the Unix Time In Minutes back to Unix Time in order to have the same condition as Unix Time
+	if len(suffix) != 10 && len(suffixStringMultiply) != 10 {
 		return ""
 	}
 
-	if !stringInRuneset(suffix, cronJobAllowedString) {
-		// Invalid suffix
+	// - Checking for unexpected character such as having characters others than numbers
+	if containUnexpectedRuneSet(suffix, cronJobUnexpectedRegex) || containUnexpectedRuneSet(suffixStringMultiply, cronJobUnexpectedRegex) {
 		return ""
 	}
 
 	return name[:lastDash]
 }
 
-func stringInRuneset(name, subset string) bool {
-	for _, r := range name {
-		if !strings.ContainsRune(subset, r) {
-			// Found an unexpected rune in suffix
-			return false
-		}
-	}
-	return true
+func containUnexpectedRuneSet(name string, unexpectedRegEx string) bool {
+	var validUnixTimeRegExp = regexp.MustCompile(unexpectedRegEx)
```

Comment:
I don't like having a `MustCompile` in here every time that this function is called, on two fronts.

1. We assume that the regex always compiles here. Otherwise, we get a hard panic if the string does not compile. 
2. We take a computational hit for the compilation every time this is called. Rather than doing this, I think you could pass in a `*regexp.Regexp` pointer, and at the top level, create and compile the regular expressions once.

Reply:
Reasonable for reason 1. However, I am having a hard time to understand the reason 2 though.

Reply:
There is a cost associated with compiling the regex. If we only call this once or twice when the agent first starts up, I think it's fine. But if this gets called during the lifecycle of the agent, then there is a CPU cost that we incur for compiling the regular expression every time. In this case, we only have two static regular expressions to choose from, so there's no need to dynamically compile the input regex string every time. We can compile the regular expressions once at the top level in a `var ( )` block, and then refer to it as needed.

Take a look at https://github.com/aws/amazon-cloudwatch-agent/blob/master/plugins/processors/k8sdecorator/stores/podstore.go#L23-L36 as an example

Reply:
Interesting. I thought it would be better to use dynamic compiling. However, you are right about if we only have two then we only need to refer when it is needed instead of complexing it though. Also interesting about agent lifecycle though.